sql/core/src/main/scala/org/apache/spark/sql/execution/python/TransformWithStateInPandasExec.scala

```diff
@@ -191,8 +223,15 @@ case class TransformWithStateInPandasExec(
     val updatesStartTimeNs = currentTimeNs
 
     val (dedupAttributes, argOffsets) = resolveArgOffsets(child.output, groupingAttributes)
-    val data =
-      groupAndProject(dataIterator, groupingAttributes, child.output, dedupAttributes)
+    // If timeout is based on event time, then filter late data based on watermark
+    val filteredIter = watermarkPredicateForDataForLateEvents match {
```

Comment:
Wow, I wonder how we could miss this... Nice catch! I was about to ask for a new test covering this, but I'll revisit once you update the test as your new test seems to also contain this part.

Reply:
Included a check on eviction of late events in the test case.